From patchwork Thu Oct  4 21:40:23 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: "Singh, Brijesh" <brijesh.singh@amd.com>
X-Patchwork-Id: 10626901
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7A9AC15E2
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  4 Oct 2018 21:40:35 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 67C3929600
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  4 Oct 2018 21:40:35 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 5B69929644; Thu,  4 Oct 2018 21:40:35 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D253729600
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  4 Oct 2018 21:40:34 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727444AbeJEEfn (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 5 Oct 2018 00:35:43 -0400
Received: from mail-eopbgr730041.outbound.protection.outlook.com
 ([40.107.73.41]:25120
        "EHLO NAM05-DM3-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1725997AbeJEEfn (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 5 Oct 2018 00:35:43 -0400
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=amdcloud.onmicrosoft.com; s=selector1-amd-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=XbYvstHbBBtFm2nKMD/RgM1zXH9mSzseubDqQzg0p2M=;
 b=0LOGyl5pA7SAflvCekY7RdAhhAE2RDyW53/QuTAAozk4o30HurFReBEZ8UWTUQ4C0hSF1oH3X1XC5TzpgIUc05ZIjP665ksxhqQi795IIvGh3jdLGt1P27B0e18uyzb0YHWBwZBwV4HQK0hGx4ADxylXBcKHjAAMbN2vrhTZkOw=
Received: from DM6PR12MB2682.namprd12.prod.outlook.com (20.176.116.31) by
 DM6PR12MB2602.namprd12.prod.outlook.com (20.176.116.11) with Microsoft SMTP
 Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id
 15.20.1185.25; Thu, 4 Oct 2018 21:40:23 +0000
Received: from DM6PR12MB2682.namprd12.prod.outlook.com
 ([fe80::7c34:f2e8:a8a9:cab8]) by DM6PR12MB2682.namprd12.prod.outlook.com
 ([fe80::7c34:f2e8:a8a9:cab8%3]) with mapi id 15.20.1185.024; Thu, 4 Oct 2018
 21:40:23 +0000
From: "Singh, Brijesh" <brijesh.singh@amd.com>
To: "linux-kernel@vger.kernel.org" <linux-kernel@vger.kernel.org>,
        "iommu@lists.linux-foundation.org" <iommu@lists.linux-foundation.org>
CC: "Singh, Brijesh" <brijesh.singh@amd.com>, "Lendacky,
 Thomas" <Thomas.Lendacky@amd.com>, Joerg Roedel <joro@8bytes.org>,
 Borislav Petkov <bp@suse.de>, Paolo Bonzini <pbonzini@redhat.com>,
	=?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 "kvm@vger.kernel.org" <kvm@vger.kernel.org>,
 Boris Ostrovsky <boris.ostrovsky@oracle.com>,
 "# 4 . 14+" <stable@vger.kernel.org>
Subject: [PATCH] iommu/amd: Clear memory encryption mask from physical address
Thread-Topic: [PATCH] iommu/amd: Clear memory encryption mask from physical
 address
Thread-Index: AQHUXCrb10G8bEZUVkePsT0raSnuLQ==
Date: Thu, 4 Oct 2018 21:40:23 +0000
Message-ID: <1538689210-14524-1-git-send-email-brijesh.singh@amd.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-clientproxiedby: SN4PR0401CA0023.namprd04.prod.outlook.com
 (2603:10b6:803:21::33) To DM6PR12MB2682.namprd12.prod.outlook.com
 (2603:10b6:5:4a::31)
authentication-results: spf=none (sender IP is )
 smtp.mailfrom=brijesh.singh@amd.com;
x-ms-exchange-messagesentrepresentingtype: 1
x-originating-ip: [165.204.77.1]
x-ms-publictraffictype: Email
x-microsoft-exchange-diagnostics: 
 1;DM6PR12MB2602;20:ZNNlULHxPPyeOBwCaXPNRjJnYrYao53whobPvqv5SBjOqFHLzu6pbYRZI9ldUBko1XqHx67DNttTKz5gHQ3koLeEZHj+v8JxUVWKxR4yllBvvpQF7Yg7YrEWfovH968uW3uLm/REPQg8elYp/GzJe2S6pacLaQ10H2LEGMOqcjHhzImnHztYrbnw70G7LkjHvXFi6tAHQI3dtplo0HRGQ1ZSH48CPv75CIZFfZVBE/Uq4HeaP4yoRoMSglU2FIQk
x-ms-office365-filtering-correlation-id: d21d1808-a1bd-48b9-d735-08d62a41fd34
x-ms-office365-filtering-ht: Tenant
x-microsoft-antispam: 
 BCL:0;PCL:0;RULEID:(7020095)(4652040)(8989299)(4534185)(4627221)(201703031133081)(201702281549075)(8990200)(5600074)(711020)(4618075)(2017052603328)(7153060)(7193020);SRVR:DM6PR12MB2602;
x-ms-traffictypediagnostic: DM6PR12MB2602:
x-microsoft-antispam-prvs: 
 <DM6PR12MB2602248DC8E5AB6DEEC755E9E5EA0@DM6PR12MB2602.namprd12.prod.outlook.com>
x-exchange-antispam-report-test: 
 UriScan:(9452136761055)(146099531331640)(767451399110);
x-ms-exchange-senderadcheck: 1
x-exchange-antispam-report-cfa-test: 
 BCL:0;PCL:0;RULEID:(8211001083)(6040522)(2401047)(5005006)(8121501046)(823302091)(10201501046)(3231355)(944501410)(52105095)(3002001)(93006095)(93001095)(6055026)(149066)(150057)(6041310)(20161123560045)(20161123562045)(201703131423095)(201702281528075)(20161123555045)(201703061421075)(201703061406153)(20161123558120)(20161123564045)(201708071742011)(7699051);SRVR:DM6PR12MB2602;BCL:0;PCL:0;RULEID:;SRVR:DM6PR12MB2602;
x-forefront-prvs: 0815F8251E
x-forefront-antispam-report: 
 SFV:NSPM;SFS:(10009020)(39860400002)(346002)(136003)(376002)(396003)(366004)(199004)(189003)(110136005)(99286004)(26005)(54906003)(81156014)(6486002)(6506007)(81166006)(68736007)(386003)(102836004)(478600001)(52116002)(8676002)(14444005)(53936002)(6512007)(71190400001)(66574009)(86362001)(575784001)(1857600001)(256004)(186003)(36756003)(2616005)(71200400001)(6116002)(2906002)(3846002)(8936002)(5660300001)(476003)(2501003)(97736004)(7736002)(5250100002)(486006)(106356001)(14454004)(2900100001)(66066001)(105586002)(316002)(25786009)(4326008)(6436002)(305945005);DIR:OUT;SFP:1101;SCL:1;SRVR:DM6PR12MB2602;H:DM6PR12MB2682.namprd12.prod.outlook.com;FPR:;SPF:None;LANG:en;PTR:InfoNoRecords;A:1;MX:1;
received-spf: None (protection.outlook.com: amd.com does not designate
 permitted sender hosts)
x-microsoft-antispam-message-info: 
 +DCuYfDFPq/A9wc/7TpI+YuhuoMwNWAHprywmEa+mzH71PW9a6Qkn8Yp+fJXFclcyiYufymskAEapKaOXK37DBjgluSy0pH+zHHfCunT1RBIkDSmvrnA2Xr+5V6HkO/naiekcw0sbUKns0R38KL0neJZGMTY7jEt9YH73a68H7d499ayVQ6Wi+e0t6LqacF/w887NLw8yacADCw7eSqxuKdVHYLcV9XShXWHHXFq0uEfpzg1pPGiJjbUwi7pMqUGCwk8F2eKr2KbfEF1BDs8a9EUMJJIYSqtFLQk7l3vFIfl6cDT2To93t2pfqT8oQaOMDcNoy/2VZQW+DOGB3hw6y4qOtUbEWDDSXDh2d+M4PY=
spamdiagnosticoutput: 1:99
spamdiagnosticmetadata: NSPM
Content-Type: text/plain; charset="utf-8"
Content-ID: <8B679516497FEB45AC5FE7EE73D29D08@namprd12.prod.outlook.com>
MIME-Version: 1.0
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 d21d1808-a1bd-48b9-d735-08d62a41fd34
X-MS-Exchange-CrossTenant-originalarrivaltime: 04 Oct 2018 21:40:23.5017
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-Transport-CrossTenantHeadersStamped: DM6PR12MB2602
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Boris Ostrovsky reported a memory leak with device passthrough when SME
is active.

The VFIO driver uses iommu_iova_to_phys() to get the physical address for
an iova. This physical address is later passed into vfio_unmap_unpin() to
unpin the memory. The vfio_unmap_unpin() uses pfn_valid() before unpinning
the memory. The pfn_valid() check was failing because encryption mask was
part of the physical address returned. This resulted in the memory not
being unpinned and therefore leaked after the guest terminates.

The memory encryption mask must be cleared from the physical address in
iommu_iova_to_phys().

Fixes: 2543a786aa25 ("iommu/amd: Allow the AMD IOMMU to work with memory encryption")
Reported-by: Boris Ostrovsky <boris.ostrovsky@oracle.com>
Cc: Tom Lendacky <thomas.lendacky@amd.com>
Cc: Joerg Roedel <joro@8bytes.org>
Cc: <iommu@lists.linux-foundation.org>
Cc: Borislav Petkov <bp@suse.de>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: kvm@vger.kernel.org
Cc: Boris Ostrovsky <boris.ostrovsky@oracle.com>
Cc: <stable@vger.kernel.org> # 4.14+
Signed-off-by: Brijesh Singh <brijesh.singh@amd.com>
---
 drivers/iommu/amd_iommu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/iommu/amd_iommu.c b/drivers/iommu/amd_iommu.c
index 73e47d9..bee0dfb 100644
--- a/drivers/iommu/amd_iommu.c
+++ b/drivers/iommu/amd_iommu.c
@@ -3069,7 +3069,7 @@ static phys_addr_t amd_iommu_iova_to_phys(struct iommu_domain *dom,
 		return 0;
 
 	offset_mask = pte_pgsize - 1;
-	__pte	    = *pte & PM_ADDR_MASK;
+	__pte	    = __sme_clr(*pte & PM_ADDR_MASK);
 
 	return (__pte & ~offset_mask) | (iova & offset_mask);
 }
